### PR TITLE
[HOLD] DEVELOPER-1792 - Docker-selenium addition

### PR DESCRIPTION
**Pull request by RubyGemTSL**
split out docker compose file. Started internal task (not complete)

-- adding wip

Adding Olly's fix for internal docker testing

-- adding wip

-- adding wip

-- completed docker-selenium addition

### CODE REVIEW

**Comment by RubyGemTSL** [comment]
@ollyjshaw - can you review this? Tests will fail as FF is not installed on CI box yet.

**Comment by ollyjshaw** [comment]
retest this please

**Comment by ollyjshaw** [comment]
retest this please

**Comment by ollyjshaw** [comment]
Can you rebase? Builds are doing funny things

**Comment by ollyjshaw** [comment]
retest this please

**Comment by ollyjshaw** [comment]
Expected failure, firefox not on build box

**Comment by paulrobinson** [comment]
@RubyGemTSL let me know what needs installing. Ideally give me a 'yum' command to run on the CI nodes.

File: _cucumber/features/support/env.rb [hunk 1]
```diff
@@ -20,47 +20,48 @@
...
+Capybara.register_driver :local_chrome do |app|
+  Capybara::Selenium::Driver.new(app, :browser => :chrome, :switches => %w[--disable-popup-blocking --ignore-ssl-errors=yes])
 end
 
+drivers = [:firefox, :local_firefox, :chrome, :local_chrome]
+drivers.each do |d|
+  Capybara::Screenshot.register_driver(d) do |driver, path|
+    driver.browser.save_screenshot path
+  end
+end
```
Comment (ollyjshaw) [inline _cucumber/features/support/env.rb]: You're registering all the drivers all the time. This is fine as long as we're sure there are no issues if chrome is not on the machine for example

File: _cucumber/features/support/hooks.rb [hunk 2]
```diff
@@ -1,14 +1,9 @@
...
   @page = App.new(@driver)
-  if Capybara.current_driver == 'poltergeist'.to_sym
-    @driver.clear_cookies
-  else
-    @driver.browser.manage.delete_all_cookies
-  end
   visit('/')
 end
-
+
```
Comment (ollyjshaw) [inline _cucumber/features/support/hooks.rb]: whitespace, 2 points to me :tongue:

File: _docker/test/test_options.rb [hunk 3]
```diff
@@ -231,6 +236,7 @@ def test_drupal_nightly
     end
 
     private def expected_unit_test_tasks
-      ['--no-deps', '--rm', 'awestruct_acceptance', 'bundle exec rake test']
-    end
+      ['--no-deps', '--rm', 'awestruct_unit_tests', 'bundle exec rake test']
+            end
```
Comment (ollyjshaw) [inline _docker/test/test_options.rb]: Whitespace

File: _docker/docker-compose.yml.erb [hunk 4]
```diff
@@ -52,6 +83,10 @@ awestruct_preview_no_reload:
     - DRUPAL_HOST_PORT=<%=ENV['DRUPAL_HOST_PORT']%>
     - DRUPAL_HOST_IP
     - HOST_TO_TEST
+    - SELENIUM_HOST
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: - HOST_TO_TEST
     - SELENIUM_HOST
I don't see why these are needed. awestruct_preview_no_reload is not responsible for testing

File: _docker/docker-compose.yml.erb [hunk 5]
```diff
@@ -92,6 +127,10 @@ awestruct:
...
+    - DRIVER
+    - PARALLEL_TEST
+
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: +    - SELENIUM_HOST
 +    - DRIVER
 +    - PARALLEL_TEST
I don't see why these are needed. awestruct is not responsible for testing

File: _docker/lib/options.rb [hunk 6]
```diff
@@ -68,28 +68,41 @@ def self.parse(args)
...
+        tasks[:acceptance_test_target_task] = ["--rm", "awestruct_acceptance_docker_internal"]
       end
 
-      opts.on('--acceptance_test_target HOST_TO_TEST [=PARALLEL_TEST]', String, 'runs the cucumber features against the specified HOST_TO_TEST') do |h, parallel='false'|
-        if h.start_with?("http://docker")
+      opts.on('--acceptance_test_target HOST_TO_TEST', String, 'runs the cucumber features against the specified HOST_TO_TEST') do |host|
+        ENV['HOST_TO_TEST'] = host
+
+        driver = (ENV['DRIVER'].to_s.empty?) ? ENV['DRIVER']='firefox' : ENV['DRIVER']
+        profile = (ENV['PARALLEL_TEST'].to_s.empty?) ? ENV['PARALLEL_TEST']='true' : ENV['PARALLEL_TEST']
```
Comment (ollyjshaw) [inline _docker/lib/options.rb]: These should be done the same in as in the --acceptance_test_docker task. Don't mid which way.

File: _docker/test/test_options.rb [hunk 7]
```diff
@@ -121,13 +121,34 @@ def test_awestruct_command
...
+        assert(tasks[:decrypt])
+        assert(tasks[:set_ports])
+        assert(tasks[:build])
+        assert_equal('firefox', ENV['DRIVER'])
+        assert_equal('true', ENV['PARALLEL_TEST'])
+        assert_equal(tasks[:supporting_services], %w(-d elasticsearch mysql searchisko searchiskoconfigure))
+        refute(tasks[:awestruct_up_service])
+        refute(tasks[:awestruct_command_args])
+        assert_equal(["--rm", "awestruct_acceptance_docker_internal"], tasks[:acceptance_test_target_task])
+      end
```
Comment (ollyjshaw) [inline _docker/test/test_options.rb]: tests look good

File: _cucumber/docker_run.sh [hunk 8]
```diff
@@ -6,4 +6,7 @@ do
   sleep 5
 done
 echo "connected successfully"
-bundle exec rake features
+echo "Executing tests using:${DRIVER} browser"
+echo "Executing tests with parallel mode set to:${PARALLEL_TEST}"
+
+bundle exec rake features DRIVER=$DRIVER PARALLEL_TEST=$PARALLEL_TEST
```
Comment (ollyjshaw) [inline _cucumber/docker_run.sh]: Can we use a different var than DRIVER. It may clash

File: _cucumber/features/support/env.rb [hunk 9]
```diff
@@ -20,47 +20,48 @@
...
+    raise 'please provide a variable for HOST_TO_TEST'
   else
     host_to_test = ENV['HOST_TO_TEST']
   end
 
   config.app_host = host_to_test
   config.run_server = false
-  config.default_driver = (ENV['DRIVER'] ||= 'poltergeist').to_sym
-  config.default_max_wait_time = 5
+  config.default_driver = (ENV['DRIVER'] ||= 'local_firefox').to_sym
```
Comment (ollyjshaw) [inline _cucumber/features/support/env.rb]: Can we use a different var than DRIVER. It may clash

File: _docker/docker-compose.yml.erb [hunk 10]
```diff
@@ -1,4 +1,30 @@
...
+  build: ./awestruct
+  command:
+     - /home/awestruct/developer.redhat.com/_cucumber/docker_run.sh
+
+  <% if ENV['ghprbPullId'] %>
+  labels:
+    - "com.develpers.redhat.pr=<%=ENV['ghprbPullId']%>"
+  <% end %>
+  links:
+    - selenium-<%=ENV['DRIVER']%>:driver
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: Can we use a different var than DRIVER. It may clash

File: _docker/docker-compose.yml.erb [hunk 11]
```diff
@@ -8,11 +34,16 @@ awestruct_acceptance:
   <% end %>
   links:
     - awestruct_preview_no_reload:docker
+    - selenium-<%=ENV['DRIVER']%>:driver
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: Can we use a different var than DRIVER. It may clash

File: _docker/docker-compose.yml.erb [hunk 12]
```diff
@@ -8,11 +34,16 @@ awestruct_acceptance:
...
   volumes:
     - ../:/home/awestruct/developer.redhat.com
   environment:
     - HOST_TO_TEST=http://docker:<%=ENV['AWESTRUCT_HOST_PORT']%>
     - AWESTRUCT_HOST_PORT=<%=ENV['AWESTRUCT_HOST_PORT']%>
+    - SELENIUM_HOST=http://driver:4444/wd/hub
+    - DRIVER
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: Can we use a different var than DRIVER. It may clash

File: _docker/docker-compose.yml.erb [hunk 13]
```diff
@@ -52,6 +83,10 @@ awestruct_preview_no_reload:
     - DRUPAL_HOST_PORT=<%=ENV['DRUPAL_HOST_PORT']%>
     - DRUPAL_HOST_IP
     - HOST_TO_TEST
+    - SELENIUM_HOST
+    - DRIVER
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: Can we use a different var than DRIVER. It may clash

File: _docker/docker-compose.yml.erb [hunk 14]
```diff
@@ -92,6 +127,10 @@ awestruct:
...
```
Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: Can we use a different var than DRIVER. It may clash

File: _docker/lib/options.rb [hunk 15]
```diff
@@ -68,28 +68,41 @@ def self.parse(args)
         tasks[:supporting_services] += %w(elasticsearch mysql searchisko searchiskoconfigure)
       end
 
-      opts.on('--acceptance_test_docker [=PARALLEL_TEST]', String, 'runs the cucumber features against the local running docker stack.') do |parallel='false'|
-        ENV['PARALLEL_TEST'] = parallel
+      opts.on('--acceptance_test_docker', String, 'runs the cucumber features against the local running docker stack.') do
+        if ENV['DRIVER'].to_s.empty?
+          ENV['DRIVER']='firefox'
```
Comment (ollyjshaw) [inline _docker/lib/options.rb]: Can we use a different var than DRIVER. It may clash

File: _docker/lib/options.rb [hunk 16]
```diff
@@ -68,28 +68,41 @@ def self.parse(args)
...
-        tasks[:acceptance_test_target_task] = ["--rm", "awestruct_acceptance", "bundle exec rake features PARALLEL_TEST=#{ENV['PARALLEL_TEST']}"]
+        tasks[:acceptance_test_target_task] = ["--rm", "awestruct_acceptance_docker_internal"]
       end
 
-      opts.on('--acceptance_test_target HOST_TO_TEST [=PARALLEL_TEST]', String, 'runs the cucumber features against the specified HOST_TO_TEST') do |h, parallel='false'|
-        if h.start_with?("http://docker")
+      opts.on('--acceptance_test_target HOST_TO_TEST', String, 'runs the cucumber features against the specified HOST_TO_TEST') do |host|
+        ENV['HOST_TO_TEST'] = host
+
+        driver = (ENV['DRIVER'].to_s.empty?) ? ENV['DRIVER']='firefox' : ENV['DRIVER']
```
Comment (ollyjshaw) [inline _docker/lib/options.rb]: Can we use a different var than DRIVER. It may clash

File: _docker/test/test_options.rb [hunk 17]
```diff
@@ -121,13 +121,34 @@ def test_awestruct_command
...
 
     def test_acceptance_test_target_task
-      tasks = Options.parse (["--acceptance_test_target=http://example.com"])
-      assert(tasks[:build])
-      assert_equal(tasks[:unit_tests], expected_unit_test_tasks)
-      assert_equal('http://example.com', ENV['HOST_TO_TEST'])
-      assert_equal('false', ENV['PARALLEL_TEST'])
-      refute(tasks[:set_ports])
-      assert_equal(["--no-deps", "--rm", "awestruct", "bundle exec rake features PARALLEL_TEST=#{ENV['PARALLEL_TEST']}"], tasks[:acceptance_test_target_task])
+      ClimateControl.modify DRIVER: 'firefox', PARALLEL_TEST: 'true' do
```
Comment (ollyjshaw) [inline _docker/test/test_options.rb]: Can we use a different var than DRIVER. It may clash

File: _docker/test/test_options.rb [hunk 18]
```diff
@@ -121,13 +121,34 @@ def test_awestruct_command
...
+
+    def test_acceptance_test_target_task_default_value
+      ClimateControl.modify HOST_TO_TEST: nil, AWESTRUCT_HOST_PORT: '32768', DRIVER: 'firefox', PARALLEL_TEST: 'true' do
+        tasks = Options.parse (["--acceptance_test_docker"])
+        assert(tasks[:kill_all])
+        assert_equal(tasks[:unit_tests], expected_unit_test_tasks)
+        assert(tasks[:decrypt])
+        assert(tasks[:set_ports])
+        assert(tasks[:build])
+        assert_equal('firefox', ENV['DRIVER'])
```
Comment (ollyjshaw) [inline _docker/test/test_options.rb]: Can we use a different var than DRIVER. It may clash

**Comment by ollyjshaw** [comment]
@paulrobinson @RubyGemTSL. It's looking good. I've a few comments, mainly around renaming DRIVER to something else like SELENIUM_DRIVER or RHD_SELENIUM_DRIVER

**Comment by RubyGemTSL** [comment]
Thanks for the review @ollyjshaw, will address comments.

@paulrobinson I've found an issue with this and it seems to be related to sharing the VPN connection with docker-selenium. When accessing a remote host such as staging tests fai as its can't access that host. Going to pair with Olly tomorrow on this.

File: _cucumber/features/support/env.rb (hunk 1 above)
Comment (RubyGemTSL) [inline _cucumber/features/support/env.rb]: This is only setting the screenshot path for each browser.

File: _docker/lib/options.rb (hunk 6 above)
Comment (RubyGemTSL) [inline _docker/lib/options.rb]: they are doing slightly different. Second one is setting a variable which is not needed it other option. This one liner looks better than changing it to:
    
    if ENV['DRIVER'].to_s.empty?
      driver = ENV['DRIVER']='firefox' 
    else 
       driver = ENV['DRIVER']
    end

**Comment by paulrobinson** [comment]
Needs a rebase

**Comment by ollyjshaw** [comment]
Can we please close this PR. It's killing our builds as the lack of a chrome driver does not fail fast.

**Pull request closed without merging**